pkg/destroy/libvirt/libvirt_prefix_deprovision.go

```diff
@@ -0,0 +1,241 @@
+package libvirt
+
+import (
+	"os"
+	"strings"
+	"time"
+
+	libvirt "github.com/libvirt/libvirt-go"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// filterFunc allows filtering based on names.
+// returns true, when the name should be handled.
+type filterFunc func(name string) bool
+
+// ClusterNamePrefixFilter returns true for name
+// that are prefixed with clustername.
+// clustername cannot be empty
+var ClusterNamePrefixFilter = func(clustername string) filterFunc {
+	if clustername == "" {
+		panic("clustername cannot be empty")
+	}
+	return func(name string) bool {
+		return strings.HasPrefix(name, clustername)
+	}
+}
+
+// AlwaysTrueFilter returns true for all
+// name expect `default`.
+var AlwaysTrueFilter = func() filterFunc {
+	return func(name string) bool {
+		if name == "default" {
+			return false
+		}
+		return true
+	}
+}
+
+// deleteFunc type is the interface a function needs to implement to be called as a goroutine.
+// The (bool, error) return type mimics wait.ExponentialBackoff where the bool indicates successful
+// completion, and the error is for unrecoverable errors.
+type deleteFunc func(conn *libvirt.Connect, filter filterFunc, logger log.FieldLogger) (bool, error)
+
+// ClusterUninstaller holds the various options for the cluster we want to delete.
+type ClusterUninstaller struct {
+	LibvirtURI string
+	Filter     filterFunc
+	Logger     log.FieldLogger
+}
+
+// Run is the entrypoint to start the uninstall process
+func (o *ClusterUninstaller) Run() error {
+	deleteFuncs := map[string]deleteFunc{}
+	populateDeleteFuncs(deleteFuncs)
+	returnChannel := make(chan string)
+
+	conn, err := libvirt.NewConnect(o.LibvirtURI)
+	if err != nil {
+		return err
+	}
+
+	// launch goroutines
+	for name, function := range deleteFuncs {
+		go deleteRunner(name, function, conn, o.Filter, o.Logger, returnChannel)
+	}
+
+	// wait for them to finish
+	for i := 0; i < len(deleteFuncs); i++ {
+		select {
+		case res := <-returnChannel:
+			o.Logger.Debugf("goroutine %v complete", res)
+		}
+	}
+
+	return nil
+}
+
+func deleteRunner(deleteFuncName string, dFunction deleteFunc, conn *libvirt.Connect, filter filterFunc, logger log.FieldLogger, channel chan string) {
+	backoffSettings := wait.Backoff{
+		Duration: time.Second * 10,
+		Factor:   1.3,
+		Steps:    100,
+	}
+
+	err := wait.ExponentialBackoff(backoffSettings, func() (bool, error) {
+		return dFunction(conn, filter, logger)
+	})
+
+	if err != nil {
+		logger.Fatalf("Unrecoverable error/timed out: %v", err)
```

Comment:
an error from one of the runner means it cannot continue. and all runners must be running to make progress.

Reply:
> an error from one of the runner means it cannot continue. and all runners must be running to make progress.

Then if the caller receives an error it can `Cancel()` the `Context`, wait out the canceled goroutines, and exit cleanly.  But I'm fine rolling with what you have and revisiting this later if you don't want to pick it up now.